### PR TITLE
docs: Add detailed document on PP Backend functionality, interactions and required technologies

### PR DESCRIPTION
**Pull request by Awambeng**
In this commit, a comprehensive document is created, providing a thorough overview of the PP Backend's functionality and its interactions with other components. The document also includes detailed information on the technologies necessary for implementing the backend and also how it should be tested. This documentation will serve as a valuable resource for understanding and developing the PP Backend effectively.

### CODE REVIEW

File: docs/backend_guide.md [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        transaction_id (Primary Key)
+        sender_id (Foreign Key referencing User Table)
+        receiver_id (Foreign Key referencing User Table)
+        amount
+        date_created
+
+This schema includes tables for users, transactions, and transaction history. The user table stores user information, including username, email, and password. The transaction table tracks payment transactions between users, including sender, receiver, and amount. The transaction history table records the status and update timestamps for each transaction. Keep in mind that this is a basic example, and the actual database schema for the Power Pay App may require additional tables and relationships. 
+
+
+# Authentication
```
Comment (stephane-segning) [inline docs/backend_guide.md:131]: Ok, this is the tricky part of our system. We want to allow users only from our domain. This means that we're not gonna allow some hackers coming from different websites. Also we'll be communicating with the translator service for exhanging SMS.

1. How can we ensure a Auth request is coming from our system?
2. How can I ensure the translator service is not compromised or fake?

This is what you should discuss here. The authentication seen by the user is cute but non-technical.

Comment (stephane-segning) [reply docs/backend_guide.md:131]: To help you understand this, you know that our system doesn't use keycloak or similar oauth2 provider for security. Meaning we should implement security measures ourselves. We should then deeply understand how to secure the PWA by itself, when it interacts with our backend, without even talking about authentication. Same goes for the Translator service.

File: docs/backend_guide.md [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            * PIN Setup: After successful OTP verification, the user is prompted to set a PIN for subsequent authentication. The PIN is securely stored in the user database, preferably using hashing and salting techniques.
+            
+        2. Choronko User Authentication:
+
+            * SMS Gateway Integration: The backend integrates with an SMS gateway to receive and interpret messages sent by Choronko users.
+            * Syntax Interpretation: The SMS gateway extracts the phone number and PIN from the incoming message using the specified syntax, such as numberpin#.
+            * PIN Verification: The backend verifies the received PIN against the one stored in the Choronko user's database record. If the PIN matches, the user is considered authenticated.
+
+
+# Error Handling
```
Comment (stephane-segning) [inline docs/backend_guide.md:149]: I wanted you to discuss these before, at the place you discussed the APIs:

1. The `POST:/api/users` should create users and return 
  - status `200` if everything was alright alongside an object containing the id of the _procedure / registration procedure_, knowing a OTP code would arrive the phone number of the user soon enough...
  - status `400` if the request contains a bad data or...
  - and so on...

File: docs/backend_guide.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+    Error handling is an essential aspect of any application, including the Power Payment Backend, to ensure smooth operation and a positive user experience. Proper error handling involves identifying, capturing, and appropriately responding to errors that may occur during the execution of backend processes. Below is an explanation of how error handling should be done in the backend, along with examples:
+
+    - HTTP Status Codes:
+
+        The app utilizes appropriate HTTP status codes to indicate the nature of the error, allowing clients to understand the response at a glance.
+
+        For example, a 400 status code can be used for client-side errors (e.g., invalid request parameters), while a 500 status code can be used for server-side errors (e.g., internal server errors).
+
+    - Error Logging and Notification:
```
Comment (stephane-segning) [inline docs/backend_guide.md:159]: Very good point

Comment (stephane-segning) [inline docs/backend_guide.md:159]: Good point. You should also indicate how this logging is done, if it's going to include another component in our system. E.g: tools like rollbar or sentry are the external tools used for logging and error handling and SHOULD be documented. Same for the notifications.

File: docs/backend_guide.md [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+    - User-Friendly Error Messages:
+
+        Error messages should be user-friendly, informative, and provide actionable instructions or suggestions whenever possible.
+
+        Example: If a user attempts to authenticate with an incorrect PIN, they can be presented with an error message stating, "Invalid PIN. Please double-check your PIN and try again."
+
+    - Graceful Degradation:
+
+        The backend should gracefully handle errors and degrade functionality when necessary to maintain system stability and prevent cascading failures.
```
Comment (stephane-segning) [inline docs/backend_guide.md:178]: You shouldn't write "should", but indicate directly how

File: docs/backend_guide.md [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+        Example: If an external API used for a specific feature is temporarily unavailable, the backend can gracefully degrade by presenting an alternative workflow or disabling the feature temporarily until the API becomes accessible again.
+
+    - Error Response Formats:
+
+        The backend should provide consistent error response formats, such as JSON or XML, that include relevant error codes, messages, and additional details for easier identification and debugging by client applications.
```
Comment (stephane-segning) [inline docs/backend_guide.md:184]: Should --> explain how

File: docs/backend_guide.md [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    - Input Validation and Sanitization:
+
+        Input data should be validated and sanitized to prevent common security vulnerabilities such as SQL injection or cross-site scripting (XSS) attacks.
+
+        Example: Prior to processing a user's input for a payment amount, the backend should validate that it is a valid numeric value and sanitize it to remove any potential harmful characters.
+
+
+# Testing
+
+    Each backend feature of the Power Pay app should undergo thorough testing using written code or test frameworks to ensure its functionality and reliability. By writing test cases and using appropriate frameworks, you can automate the testing process and identify potential issues early on. Here are some examples of how different backend features can be tested:
```
Comment (stephane-segning) [inline docs/backend_guide.md:197]: In this part, you should be discussing how we do tests in the backend. How we do unit testing and how we do integration testing. And out that in relationship with the components in the architecture of the backend, that you skipped. Not processes, it's not helpful

**Review by stephane-segning — CHANGES_REQUESTED** [review]
A lot can be done better. Please understand these request for changes and work over it.

File: docs/backend_guide.md [hunk 7]
```diff
@@ -0,0 +1,227 @@
+# Project Overview
+
+The Power Payment Backend is a cutting-edge system designed to facilitate seamless money transfers between users through a Progressive Web App (PWA). The primary purpose of the project is to provide a secure and user-friendly platform for users to send and receive money electronically.
```
[the remark this answers is not in the document]
Comment (Awambeng) [reply docs/backend_guide.md:3]: The reason why i added this was to give a small overview to create awareness for those who will be working at the backend but if i understand well your requirement you want me to remove this?

File: docs/backend_guide.md [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+The target audience for this project includes individuals, businesses, and organizations that require a reliable and efficient payment solution. Whether it's friends splitting bills, businesses paying suppliers, or organizations distributing funds, the Power Payment Backend caters to a wide range of users.
```
[the remark this answers is not in the document]
Comment (Awambeng) [reply docs/backend_guide.md:5]: and this too

File: docs/backend_guide.md (hunk 6 above)
Comment (Awambeng) [reply docs/backend_guide.md:197]: okay

File: docs/backend_guide.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (stephane-segning) [reply docs/backend_guide.md]: Yes. Thos who'll read these will be having all other description, so they'll know. The global `README.md` is also going to introduce all this.

File: docs/image.png
Comment (stephane-segning) [inline docs/image.png:1]: Remove this image. Do your graphs using mermaid only

File: docs/backend_guide.md [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+                        }
+                    }```
+
+            2. Centralized Logging System:
+
+            Configure the Power Payment Backend to use a centralized logging system, such as ELK Stack (Elasticsearch, Logstash, Kibana). This allows for the aggregation, analysis, and visualization of logs from different components of the backend.
+
+            3. Error Notifications:
+
+            Utilize a notification service or tool to trigger alerts when critical errors occur. For example, you can configure an email notification to be sent to the development team or a dedicated Slack channel. This ensures the right people are promptly informed of critical issues. Here's an example using Spring Boot's email notification:
```
Comment (stephane-segning) [inline docs/backend_guide.md:182]: Why do you write code? The change I was expecting was way to notify errors. What mechanisms do we have? Which one do we want to have? Is email better? Who where is going to be the Smtp server?

Comment (stephane-segning) [reply docs/backend_guide.md:182]: Also, isn't it faster and cost effective to use a pre-existing solution?

File: docs/backend_guide.md (hunk 4 above)
Comment (stephane-segning) [reply docs/backend_guide.md:253]: I meant for you to explain how, not to write code.

File: docs/backend_guide.md (hunk 5 above)
Comment (stephane-segning) [reply docs/backend_guide.md:288]: Same

File: docs/backend_guide.md (hunk 6 above)
Comment (stephane-segning) [reply docs/backend_guide.md:329]: I find this one not good. Try thinking about technologies, not about how scenarios would be tested. That way, you'll have a list of technologies and a list of scenarios. Then build a matrix of scenarios and each their testing libraries.

**Review by stephane-segning — CHANGES_REQUESTED** [review]
Please work according to the points I mentioned earlier and the new ones

**Comment by Awambeng** [comment]
prof @stephane-segning i think i have touched all the sections you asked me to review and correct you can now check

File: docs/backend_guide.md
Comment (stephane-segning) [inline docs/backend_guide.md:1]: Please fix the formatting of your document

**Review by stephane-segning — CHANGES_REQUESTED** [review]

File: docs/backend_guide.md [hunk 10]
```diff
@@ -0,0 +1,420 @@
...
+G[Tchoronko] --> |Auth Service| H[Telecom Provider]
+H --> B
+H --> |SMS| G
+```
+
+# System Components
+
+1. Power Pay Service: 
+
+This is the core of the backend which intelligently processes all requests and orchestrates the system's intricate logic. This service intuitively interacts with the Power Pay App, SMS Gateway, and the Power Pay Store via RESTful APIs, ensuring smooth and reliable communication channels.
```
Comment (stephane-segning) [inline docs/backend_guide.md:33]: The backend does not interact with the SMS Gateway, remove that please.

**Comment by Awambeng** [comment]
> Good work @Awambeng

Thanks prof but there are 3 workflows to be checked before i am able to merge

**Review by stephane-segning — CHANGES_REQUESTED** [review]
Good work

**Review by stephane-segning — APPROVED** [review]
Good first version!

**Pull request merged**